src/uuid.cr

```diff
@@ -0,0 +1,209 @@
+        string_has_hex_pair_at! value, offset
+        bytes[i] = value[offset, 2].to_u8(16)
+      end
+    else
+      raise ArgumentError.new "Invalid string length #{value.size} for UUID, expected 32 (hexstring), 36 (hyphenated) or 46 (urn)"
+    end
+
+    new(bytes, variant, version)
+  end
+
+  # Raises `ArgumentError` if string `value` at index `i` doesn't contain hex
+  # digit followed by another hex digit.
+  private def self.string_has_hex_pair_at!(value : String, i)
+    unless value[i, 2].to_u8(16, whitespace: false, underscore: false, prefix: false)
+      raise ArgumentError.new [
+        "Invalid hex character at position #{i * 2} or #{i * 2 + 1}",
+        "expected '0' to '9', 'a' to 'f' or 'A' to 'F'",
+      ].join(", ")
+    end
+  end
+
+  # Generates RFC 4122 v4 UUID.
+  #
+  # It is strongly recommended to use a cryptographically random source for
+  # *random*, such as `Random::Secure`.
+  def self.random(random = Random::Secure, variant = Variant::RFC4122, version = Version::V4)
+    new_bytes = uninitialized UInt8[16]
+    random.random_bytes(new_bytes.to_slice)
+
+    new(new_bytes, variant, version)
+  end
+
+  def self.empty
+    new(StaticArray(UInt8, 16).new(0_u8), UUID::Variant::NCS, UUID::Version::V4)
+  end
+
+  # Returns UUID variant.
+  def variant
+    case
+    when @bytes[8] & 0x80 == 0x00
+      Variant::NCS
+    when @bytes[8] & 0xc0 == 0x80
+      Variant::RFC4122
+    when @bytes[8] & 0xe0 == 0xc0
+      Variant::Microsoft
+    when @bytes[8] & 0xe0 == 0xe0
+      Variant::Future
+    else
+      Variant::Unknown
+    end
+  end
+
+  # Returns version based on RFC4122 format. See also `#variant`.
+  def version
+    case @bytes[6] >> 4
+    when 1 then Version::V1
+    when 2 then Version::V2
+    when 3 then Version::V3
+    when 4 then Version::V4
+    when 5 then Version::V5
+    else        Version::Unknown
+    end
+  end
+
+  # Returns 16-byte slice.
+  def to_slice
+    @bytes.to_slice
+  end
+
+  # Returns unsafe pointer to 16-bytes.
+  def to_unsafe
+    @bytes.to_unsafe
+  end
+
+  # Returns `true` if `other` UUID represents the same UUID, `false` otherwise.
+  def ==(other : UUID)
+    to_slice == other.to_slice
+  end
+
+  def to_s(io : IO)
+    slice = to_slice
+
+    buffer = uninitialized UInt8[36]
+    buffer_ptr = buffer.to_unsafe
+
+    buffer_ptr[8] = buffer_ptr[13] = buffer_ptr[18] = buffer_ptr[23] = '-'.ord.to_u8
+    slice[0, 4].hexstring(buffer_ptr + 0)
+    slice[4, 2].hexstring(buffer_ptr + 9)
+    slice[6, 2].hexstring(buffer_ptr + 14)
+    slice[8, 2].hexstring(buffer_ptr + 19)
+    slice[10, 6].hexstring(buffer_ptr + 24)
+
+    io.write(buffer.to_slice)
+  end
+
+  def hexstring
+    to_slice.hexstring
+  end
+
+  def urn
+    String.build(45) do |str|
+      str << "urn:uuid:"
+      to_s(str)
+    end
+  end
+
+  {% for v in %w(1 2 3 4 5) %}
+    # Returns `true` if UUID looks is a V{{ v.id }}, `false` otherwise.
```

Comment:
Did you mean "if UUID looks *as*"?

Reply:
"If UUID has version Vx" would make more sense surely?